pkg/api/v1/types.go

```diff
@@ -2108,7 +2108,41 @@ type NodeStatus struct {
 	// More info: http://releases.k8s.io/HEAD/docs/admin/node.md#node-info
 	NodeInfo NodeSystemInfo `json:"nodeInfo,omitempty"`
 	// List of container images on this node
-	Images []ContainerImage `json:"images,omitempty"`
+	Images []ContainerImage `json:"images",omitempty`
+	// Bounded-sized list of signatures of pods that should avoid this node, sorted
+	// in timestamp order from oldest to newest. Size of the slice is unspecified.
+	PreferAvoidPods []PreferAvoidPodsEntry `json:"preferAvoidPods,omitempty"`
+}
+
+// Describes a class of pods that should avoid this node.
+type PreferAvoidPodsEntry struct {
+	// The class of pods.
+	PodSignature PodSignature `json:"podSignature"`
+	// Time at which this entry was added to the list.
+	EvictionTime unversioned.Time `json:"evictionTime,omitempty"`
+	// (brief) reason for this entry was added to the list.
+	Reason string `json:"reason,omitempty"`
+	// Human readable message indicating why this entry was added to the list.
+	Message string `json:"message,omitempty"`
+}
+
+// Describes the class of pods that should avoid this node.
+// Exactly one field should be set.
+type PodSignature struct {
+	// Reference to controller whose pods should avoid this node.
+	ControllerRef *ControllerRef `json:"controllerRef"`
+}
+
+// ControllerRef contains enough information to let you inspect or modify the referred controller.
+type ControllerRef struct {
```

Comment:
My thoughts about ControllerRef:
- controllerRef in my opinion is supposed to be a generic concept for all objects (as an example, controllerRef of ReplicaSet can be set to deployment).
- we shouldn't reinvent the way of referencing from scratch - we already have "ObjectReference" object defined and we should use it here.